### PR TITLE
Apply adjustments to bring taiko conversions in-line with stable

### LINKED ISSUE #2141: Taiko doesn't correctly convert drum rolls for some beatmaps
On the following beatmap: https://osu.ppy.sh/p/beatmap?b=1021024&m=1 , some of the hitobjects are supposed to be drumrolls. This causes user replays to miss.

### PR DESCRIPTION
**Pull request by smoogipoo**
Fixes https://github.com/ppy/osu/issues/2141

I think I'm ready to call this done. There's two parts to this that I tackled at the same time:
1. A few minor implementation details regarding strain - stable doesn't factor in the rate.
2. Fixes to make sure that conversions match stable.

In order to fix (2), I had to copy osu!stable's calculations, which add some FP errors. Hopefully I've documented this well enough.

## Converted beatmaps

Full list of **ranked/loved convert** beatmaps that differ in SR by more than 0.01* (filtered to mods=0, from the May [dump](https://data.ppy.sh)): https://docs.google.com/spreadsheets/d/1Q9PJpBphcnL8pWxlrRIaBNw8UGDw-duo6JBN9uBZJBw/edit?usp=sharing

All differences are due to osu!lazer being more accurate with start times causing diffcalc to sometimes goes down different code paths due to minute differences in a calculation: https://github.com/ppy/osu/blob/f0fa85960913b88e96495b3fc8e1799271701690/osu.Game.Rulesets.Taiko/Difficulty/Skills/Strain.cs#L63-L65
This is fine, since the differences are minor and the code will be completely changed soon anyway. 

Out of all converted beatmaps...
```
0.07% of beatmaps have an SR delta >0.1*
0.01% of beatmaps have an SR delta >0.5*
```

## Taiko-specific beatmaps

There is only one difference in taiko-specific beatmaps: https://osu.ppy.sh/beatmapsets/654033#taiko/1386414, which is an aspire map that has a simultaneous hitobject at some point. The difference here is ordering of rim vs centre hit and doesn't fall under the above conversion issues. The SR difference for this is 0.02*.

Out of all taiko-specific beatmaps...
```
0.008% of beatmaps have an SR delta >0.01*
    0% of beatmaps have an SR delta >0.1*
```

## Testing

This is quite hard to test. Both added tests fail against current master, but they don't fail as much as they should, so there's other timing errors involved that are hard to replicate in an isolated manner - you'd really need the whole beatmaps to replicate. The test cases were built from the following beatmaps:
https://osu.ppy.sh/b/37736
https://osu.ppy.sh/b/131766
https://osu.ppy.sh/b/674909
https://osu.ppy.sh/b/1659962
I'd say what I currently have there is probably fine enough though, along with the very extensive testing I've done against all ranked/loved beatmaps above.

### CODE REVIEW

**Comment by smoogipoo** [comment]
I've also tested these changes against osu! on the same beatmaps and there's no differences.

**Comment by bdach** [comment]
I've visually checked the four maps on lazer with this PR and on stable (side-by-side) and they look to be matching. The only exception is that on https://osu.ppy.sh/b/37736 it seems like the drum roll at ~0:49 has double the tick rate on stable as compared to lazer but I'm not sure whether that has anything to do with conversion or is even relevant to diffcalc. Otherwise no major discrepancies as far as I can tell.

For obvious reasons I think this is about as much as I can check. I've looked through the code and I *think* I found the floating-point inaccuracies from stable which affect conversion, namely:

* multiplying by `BpmMultiplier` instead of dividing by `SpeedMultiplier` (which is its inverse, but dividing is better as it avoids precision loss)
* dividing by `SliderTickRate` and then multiplying by it again,
* calculating `osuVelocity` by scaling `taikoVelocity` by `1000f / beatLength` and then undoing that by multiplaying by `1000f` in the return value.

Not sure whether it's worth leaving an inline comment near each one of those to emphasise where the necessary legacy brokenness lies.

File: osu.Game.Rulesets.Taiko/Difficulty/Skills/Strain.cs [hunk 1]
```diff
@@ -26,7 +26,7 @@ protected override double StrainValueOf(DifficultyHitObject current)
             double addition = 1;
 
             // We get an extra addition if we are not a slider or spinner
-            if (current.LastObject is Hit && current.BaseObject is Hit && current.DeltaTime < 1000)
+            if (current.LastObject is Hit && current.BaseObject is Hit && (current.BaseObject.StartTime - current.LastObject.StartTime) < 1000)
```
[the remark this answers is not in the document]
Comment (smoogipoo) [reply osu.Game.Rulesets.Taiko/Difficulty/Skills/Strain.cs]: I think I prefer it this way, but sure. Keep in mind this code is going to go away very soon.

**Review by peppy — APPROVED** [review]
nice job

**Pull request merged**